### ISSUE TITLE
[Bug] Crash @kotlin.KotlinNullPointerException: at mozilla.components.service.fxa.manager.FxaAccountManager.postAuthenticated(FxaAccountManager.kt:36)

### ISSUE THREAD

**Issue by grigoryk** [mozilla-mobile/android-components#7536]
Filed from https://github.com/mozilla-mobile/fenix/issues/12000

https://crash-stats.mozilla.org/signature/?product=Fenix&signature=kotlin.KotlinNullPointerException%3A%20at%20mozilla.components.service.fxa.manager.FxaAccountManager.postAuthenticated%28FxaAccountManager.kt%3A36%29

```
kotlin.KotlinNullPointerException
	at mozilla.components.service.fxa.manager.FxaAccountManager.postAuthenticated(FxaAccountManager.kt:36)
	at mozilla.components.service.fxa.manager.FxaAccountManager$postAuthenticated$1.invokeSuspend(FxaAccountManager.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:2)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:19)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1115)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:590)
	at java.lang.Thread.run(Thread.java:818)
```

**Comment by grigoryk** [comment]
The obvious thing that can NPE in that stacktrace (unfortunately line numbers are wrong) is fxaDeviceId = account.getCurrentDeviceId()!!.

**Comment by grigoryk** [comment]
`getCurrentDeviceId` ends up reading from local state in https://github.com/mozilla/application-services/blob/master/components/fxa-client/src/device.rs#L334-L339:
```
match self.state.current_device_id {
            Some(ref device_id) => Ok(device_id.to_string()),
            None => Err(ErrorKind::NoCurrentDeviceId.into()),
        }
```

`self.state.current_device_id` is set in two places - `ensure_capabilities` or `initialize_device` (in fxaclient's device.rs). In the most common scenario (just logging-in, either manually or via pairing) we call initialize_device just before going into `postAuthenticated`. But, that needs to talk to network, and can fail.
Currently, we _assume_ that this call actually succeeded. If it didn't, we won't have the `current_device_id` set, and will blow up in `postAuthenticated`.

**Comment by pocmo** [comment]
According to Google Play this is our top crash in 79.0.0.

**Comment by pocmo** [comment]
The following Fenix issue has STRs:
https://github.com/mozilla-mobile/fenix/issues/13064

**Comment by Diana-Rus** [comment]
Hi, issue is still reproducible with Samsung Galaxy S9 (Android 8) and Google Pixel 3XL (Android 9) on Firefox Release 79.0.1.

**As long as you leave the "Approval now required" page, after a second or two and before on desktop is displayed "Firefox Account is connected with Firefox for Android."**

Crash link - Samsung Galaxy S9 (Android 8): https://crash-stats.mozilla.org/report/index/25257b31-c0f6-49ec-9892-2b4df0200731
Crash link - google Pixel 3 XL (Android 9): https://crash-stats.mozilla.org/report/index/b4d6be88-4130-4bf7-96cb-e4b8d0200731

Same STR, exception is the timing in step 6:
1. Open browser go to Settings/ Sign in with a QR code.
2. Tap on the "Allow" button from the dialog box.
3. On Firefox Nightly desktop go to firefox.com/pair and tap on the "Show code" button.
4. Scan the provided QR code.
5. Confirm pairing on both mobile and desktop.
6. Leave "Approval now required" page by tapping on deivce back button, after 1 or 2 seconds.
7. Go to Settings/Synced tabs.
8. Wait a couple of seconds.

**Comment by grigoryk** [comment]
Thanks, @Diana-Rus. These STRs are quite different now, and the stacktrace is different. Before we were dealing with error handling around network failures, and this is something else now. I've filed https://github.com/mozilla-mobile/android-components/issues/7944 to investigate. Thanks for reporting :)

**Comment by grigoryk** [comment]
@Diana-Rus landed a fix in https://github.com/mozilla-mobile/android-components/issues/7944 - I'll close this issue for now.

We've cut 48.0.12 with this fix in it, which should be available in 79.0.2 fenix release that will be available later today.
